--- a/kcm/cache.c
+++ b/kcm/cache.c
@@ -102,7 +102,7 @@ kcm_ccache_resolve_by_uuid(krb5_context
     for (p = ccache_head; p != NULL; p = p->next) {
 	if ((p->flags & KCM_FLAGS_VALID) == 0)
 	    continue;
-	if (memcmp(p->uuid, uuid, sizeof(uuid)) == 0) {
+	if (memcmp(p->uuid, uuid, sizeof(*uuid)) == 0) {
 	    ret = 0;
 	    break;
 	}
--- a/lib/hcrypto/engine.c
+++ b/lib/hcrypto/engine.c
@@ -88,7 +88,7 @@ ENGINE_finish(ENGINE *engine)
     if(engine->destroy)
 	(*engine->destroy)(engine);
 
-    memset(engine, 0, sizeof(engine));
+    memset(engine, 0, sizeof(*engine));
     engine->references = -1;
 
 
--- a/lib/hcrypto/md2.c
+++ b/lib/hcrypto/md2.c
@@ -130,5 +130,5 @@ MD2_Final (void *res, struct md2 *m)
     MD2_Update(m, pad, 16);
 
     memcpy(res, m->state, MD2_DIGEST_LENGTH);
-    memset(m, 0, sizeof(m));
+    memset(m, 0, sizeof(*m));
 }
--- a/lib/hcrypto/test_rand.c
+++ b/lib/hcrypto/test_rand.c
@@ -164,7 +164,7 @@ main(int argc, char **argv)
 
 	for (bit = 0; bit < 8; bit++) {
 
-	    res = ((double)abs(len - bits[bit] * 2)) / (double)len;
+	    res = abs(len - bits[bit] * 2) / (double)len;
 	    if (res > 0.005)
 		errx(1, "head%d vs tail%d > 0.5%%%% %lf == %d vs %d",
 		     bit, bit, res, len, bits[bit]);
--- a/configure
+++ b/configure
@@ -14242,7 +14242,7 @@ case "$host" in
 	;;
 *-*-linux* | *-*-linux-gnu)
 	case `uname -r` in
-	2.*|3.*)
+	2.*|3.*|4.*)
 		native_pthread_support=yes
 		PTHREAD_CFLAGS=-pthread
 		PTHREAD_LIBADD=-pthread
--- a/lib/asn1/gen.c
+++ b/lib/asn1/gen.c
@@ -292,7 +292,7 @@ close_generate (void)
         fclose (templatefile);
     if (logfile)
         fprintf (logfile, "\n");
-        fclose (logfile);
+    fclose (logfile);
 }
 
 void
--- a/appl/su/su.c
+++ b/appl/su/su.c
@@ -374,7 +374,7 @@ main(int argc, char **argv)
     if (su_info == NULL)
 	errx (1, "malloc: out of memory");
 
-	pwd = getpwuid(getuid());
+    pwd = getpwuid(getuid());
     if(pwd == NULL)
 	errx(1, "who are you?");
     login_info = dup_info(pwd);
--- a/lib/krb5/krbhst.c
+++ b/lib/krb5/krbhst.c
@@ -325,7 +325,7 @@ krb5_krbhst_format_string(krb5_context c
 			  char *hostname, size_t hostlen)
 {
     const char *proto = "";
-    char portstr[7] = "";
+    char portstr[12] = "";
     if(host->proto == KRB5_KRBHST_TCP)
 	proto = "tcp/";
     else if(host->proto == KRB5_KRBHST_HTTP)
